keps/sig-api-machinery/3037-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,590 @@
+- Work outside of the cluster network.
+- Provide a client-side load balancing solution to distribute the load between the available API servers.
+
+## Proposal
+
+The proposal is to extend client-go to allow to connect to multiple control-plane nodes, and for API servers to announce the list of available servers to the clients.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Kubernetes admin I want to deploy my cluster in HA without having to use external load balancers for the cluster communication.
+
+#### Story 2
+
+As a Kubernetes developer I'd like to be able to develop HA applications and controllers without requiring external components.
+
+#### Story 3
+
+As a Kubernetes architect I'd like the cluster components to be completely autonomous.
+
+### Notes/Constraints/Caveats (Optional)
+
+Network load balancing is complicated, there are many edge cases and there is no universal solution, the requirements also change and evolve with the time (see HTTP1.1 vs HTTP2).
+
+Client-go is the library used by all the cluster components to communicate with the API servers, any change or regression there can have a great impact on the project.
+
+The implementation of this feature should not modify any of current behaviors and be completely pluggable, so users can opt-in safely and the project can be sure there will not create any regression.
+
+### Risks and Mitigations
+
+The feature will be opt-in, users can disable it and keep having the same behavior.
+
+To avoid possible security issues, this feature will require HTTP2 and HTTPS in order to work.
+
+The alternative services IPs exposed are the same published in the Endpoints for the `kubernetes.default` services, the Alt-Svc headers will not be present on requests
+that are not authenticated.
+
+## Design Details
+
+There are two problems we have to solve:
+
+- get a list of available control-plane nodes
+- client capability to be able to connect to multiple control-planes
+
+### List of available control-plane nodes
+
+This problem can be solved in two different ways:
+
+1. Configuring on the client the list of available control-plane nodes
+
+This can be done extending the client-go configuration with an additional field that allows users to include a list of alternative control-plane nodes that the client can use.
+
+`k8s.io/client-go/rest/config.go`
+```go
+type Config struct {
+	// Host must be a host string, a host:port pair, or a URL to the base of the apiserver.
+	// If a URL is given then the (optional) Path of that URL represents a prefix that must
+	// be appended to all request URIs used to access the apiserver. This allows a frontend
+	// proxy to easily relocate all of the apiserver endpoints.
+	Host string
+  // AlternativeHosts must be a comma separated list if hosts, host:port pairs or URLs to the base of
+  // different apiservers. The client can use any of them to access the apiserver.
+	AlternativeHosts string
+```
+
+2. RFC7838 HTTP Alternative Services
```

Comment:
If you want to do it this way, I don't understand why we'd want to add a header to all requests instead of e.g. letting the client do a specific call to read the endpoints object.

Reply:
That was the main pushback I've received, the first prototypes used that approach , but I couldn't argue against the fact it duplicate requests on the clients

Reply:
Adding a header to every single HTTP response seems like maybe not the best way to avoid a single extra call on an empty cache.

Reply:
As I've said, my first internal proposal was that and was dropped on that argument, this is not my expertise area so one of the reasons to write the kep was to have this feedback, the additional requests instead of the header is a feasible alternative easy to implement

Reply:
Adding a header is fairly straightforward to implement server side and consume client side and a pretty low effort change. Having a new API endpoint with a new `Kind` will undoubtedly attract a lot of bike shedding. `Alt-Svc` follows existing well defined RFC, which was designed for the same problem this KEP tries to solve, why reinvent something?

Reply:
I don't think the counter-proposal is a new Kind.  We alreaady have a Kind (two actually!) that carries a list of IPs for a "service".

The reason I see for not wanting to do it that way is that clients either a) poll `.../namespaces/default/endpoints/kubernetes` (more complicated with EndpointSlices) or they have to set up a watch.

Delivering it on every API call is "easier" but doesn't fundamentaally solve the problems polling has - if you do not get an update often enough, you could have bad information.

And this still requires at least 1 "seed" IP that doesn't change, so you can start the poll/watch